[waseemingly] iP

waseemingly opened the pull request:
# Duke :computer:
“We always need little reminders that it's gonna be all good.” – Andy Grammer ([source](https://www.brainyquote.com/quotes/andy_grammer_870566?src=t_reminders))

Duke is your daily task reminder. It's,
- text-based
- ~~EASY~~ _SUPER_ EASY to learn
- quick to use

All you need to do is, 
1. download it from [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/446).
2. add your tasks, deadlines, events
3. allow it to manage it 😉

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events 
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public void run() {
        Scanner scanner = new Scanner(System.in);

        while (true) {
            try {
                String userInput = scanner.nextLine();
                Command command = Parser.parse(userInput); // This interprets the user input and returns the respective command

                command.execute(tasks, ui, storage); // Execute the command

                if (command instanceof ByeCommand) {
                    break;
                }
            } catch (DukeException e) {
                ui.showError(e.getMessage());
            } catch (IOException e) {
                ui.showSavingError(e.getMessage());
            }
        }
    }

public static void main(String[] args) throws FileNotFoundException {
        new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

waseemingly committed:
Level-0 :Rename, Greet, Exit

waseemingly committed:
Level 1. Echo
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

waseemingly committed:
Level 2. Add, List
Add the ability to store whatever text entered by the user and display them back to the user when requested.

waseemingly committed:
Level 3. Mark as Done
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

Created a new class called Task to represent tasks.

waseemingly committed:
Level 4. ToDos, Events, Deadlines
Add support for tracking three types of tasks

Created the following classes which are subclasses of Task
ToDos: tasks without any date/time attached to it
Deadlines: tasks that need to be done before a specific date/time
Events: tasks that start at a specific date/time and ends at a specific date/time

waseemingly committed:
Duke A-TextUiTesting: Automated Text UI Testing
Test using the I/O redirection technique
Use the input/output redirection technique to semi-automate the testing of Duke.

waseemingly committed:
Level 5. Handle Errors
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

Defined a new class DukeException to represent exceptions specific to Duke.

waseemingly committed:
Level 6. Delete
Add support for deleting tasks from the list.

Used java collection classes to use ArrayList<Task> to store the tasks.

waseemingly committed:
Add save functionality

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when the chatbot starts up.

waseemingly committed:
Add date and time reading functionality.

Store deadline dates as a java.time.LocalDateTime in your task objects. Accept dates in a format such as yyyy-mm-dd format (e.g., 2019-10-15) and print in a different format such as MMM dd yyyy e.g., (Oct 15 2019).

waseemingly committed:
Add more OOP and create new classes

Classes UI, Parser, TaskList, Storage, Command was added

waseemingly committed:
Remove utils.java

waseemingly committed:
remove duke.txt

waseemingly committed:
Divide classes into package duke

waseemingly committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

* 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip:
  Bump gradle and lib version
  Add Gradle support

waseemingly committed:
Add JUnit tests

waseemingly committed:
Package the app as a Jar file

waseemingly committed:
Add JavaDoc comments

waseemingly committed:
Follow the coding standard

waseemingly committed:
Add a feature where user find a task by searching for a keyword

waseemingly committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

# Conflicts:
#	src/main/java/duke/Parser.java

waseemingly committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add a feature where user find a task by searching for a keyword

cheeggered commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public Duke(String filePath) throws FileNotFoundException {
+        ui = new Ui();
+        ui.showWelcome();
```
Any reason why you put this line in the constructor instead of the run() method body?

cheeggered commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+     * @throws DukeException If the user input is not recognized or improperly formatted.
+     */
+    public static Command parse(String userInput) throws DukeException {
+        if ("bye".equalsIgnoreCase(userInput)) {
```
good use of equalsIgnoreCase method!

cheeggered commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+                String line = scanner.nextLine();
+                String[] parts = line.split(" \\| ");
+
+                switch (parts[0]) {
```
Good use of switch for neater code!

waseemingly committed:
Add checkStyle to detect coding style violations.

MadLamprey commented on gradle/wrapper/gradle-wrapper.jar:
Since this is a binary file, maybe could ignore this from tracking?

MadLamprey commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+    public static Command parse(String userInput) throws DukeException {
+        if ("bye".equalsIgnoreCase(userInput)) {
+            return new ByeCommand();
+        } else if ("list".equalsIgnoreCase(userInput)) {
```
Could make use of switch to eliminate multiple "else if"

MadLamprey reviewed (commented):
LGTM!

H1410101 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+     * @param filePath The path to the file where tasks are saved and loaded.
+     * @throws FileNotFoundException If the file is not found.
+     */
+    public Duke(String filePath) throws FileNotFoundException {
+        ui = new Ui();
+        ui.showWelcome();
+        storage = new Storage(filePath);
+        tasks = new TaskList(storage.loadTasks());
+    }
```
It seems to me that while well-handled, the bot does not start if the file has not already been created (and the only way out might be to manually create a starting file). Is this a deliberate design decision?

H1410101 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+    /**
+     * Constructs a new DukeException with the specified detail message.
+     *
+     * @param message The detail message. The detail message is saved for later retrieval by the {@link #getMessage()} method.
```
I love this `@link`. Just thought I'd mention it.

H1410101 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+        } else {
+            throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
```
To add on to the two previous reviewers' comments, and tie them together:

I like the fact that you ignore case. However, I strongly believe a switch statement here would be far easier to read.
Perhaps you could use `switch(userInput.toLowerCase())`?

H1410101 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     *
+     * @param index The index of the task to be deleted.
+     */
+    public void delete(int index) {
```
Nit: perhaps return the Task? seems like behaviour that makes sense, even though I do think IntelliJ will draw a squiggly.

H1410101 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+            }
+
+            Task task = tasks.get(taskNumber - 1);
+            if (mark) {
+                task.markAsDone();
+                System.out.println("____________________________________________________________");
+                System.out.println("Nice! I've marked this task as done:");
+            } else {
+                task.unmarkAsDone();
+                System.out.println("____________________________________________________________");
+                System.out.println("OK, I've marked this task as not done yet:");
+            }
+            System.out.println("  [" + task.getStatusIcon() + "] " + task.description);
+            System.out.println("____________________________________________________________");
+        } catch (NumberFormatException e) {
+            System.out.println("Please provide a valid task number.");
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("That task doesn't exist!");
+        }
```
While it's good that this handles its own exceptions, I do think that UI should be handled by `Ui.java` for maximal OOP-ness; while it's troublesome, I hope it's a good time to do so regardless since there's a shift to GUI anyway.

H1410101 reviewed (approved):
LGTM
I like your choice of method names!
Apart from some functionality/design nits, I think your code is great.

waseemingly committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

waseemingly committed:
Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

waseemingly committed:
Add assert feature to Parser, TaskList and Ui classes.

- Ensure non-null input for task-related methods.
- Validate task index range for access and modification methods.
- Ensure user input is not null in processTask.
- Ensure keyword is non-empty in findTasksByKeyword.
- Ensure a task number is provided after task inputs in Parser
- Ensure task methods in Ui class cannot be null and the number
  is non-negative

waseemingly committed:
Improve code quality

- Structure code logically
- Delete dead code
- Minimise scope of variables

waseemingly committed:
Merge pull request #1 from waseemingly/branch-A-Assertions

Add assert feature to Parser, TaskList and Ui classes.

waseemingly committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assert feature to Parser, TaskList and Ui classes.

waseemingly committed:
Merge pull request #2 from waseemingly/branch-A-CodeQuality

Improve code quality

waseemingly committed:
Add more flexibility in searching items.

Users can now find items even if the keyword matches the item only
partially.

waseemingly committed:
Add feature list to README.md to update product website

waseemingly committed:
Add new images to GUI

waseemingly committed:
Add a representative screenshot of the product

waseemingly committed:
no message

waseemingly committed:
Remove _config.yml file

waseemingly committed:
Update README.md

waseemingly committed:
Rename README.md to index.md

waseemingly committed:
Update index.md

waseemingly committed:
Add better GUI functionality

Dialogbox is seen clearer and can display more than 4 tasks

waseemingly committed:
Merge commit '82e7c1663969da2bb20f1696c2c4da924cf98005'

* commit '82e7c1663969da2bb20f1696c2c4da924cf98005':
  Update index.md
  Rename README.md to index.md
  Update README.md

waseemingly committed:
Update index.md

waseemingly committed:
Update index.md
